simtools/simtel/simtel_events.py

```diff
@@ -1,194 +1,253 @@
-                        'EMPTY',
-                        horizontalalignment='center',
-                        verticalalignment='center',
-                        transform=ax.transAxes
-                    )
-                    continue
+        Parameters
+        ----------
+        energyRange: Tuple (len 2)
+            Max and min energy of energy range, e.g. energyRange=(100 * u.GeV, 10 * u.TeV)
+        coreMax: astropy.Quantity (distance)
+            Maximum core radius for selecting showers, e.g. coreMax=1000 * u.m
+
+        Returns
+        -------
+        int
+            Number of simulated events.
+        '''
+        energyRange = self._validateEnergyRange(energyRange)
+        coreMax = self._validateCoreMax(coreMax)
 
-                x_bins = _get_bins(hist, axis=0)
-                y_bins = _get_bins(hist, axis=1)
+        # energy factor
+        def integral(erange):
+            power = self._mcHeader['spectral_index'] + 1
+            return math.pow(erange[0], power) - math.pow(erange[1], power)
 
-                ax.pcolormesh(x_bins, y_bins, hist['data'])
-                ax.set_xlim(xlim)
-                ax.set_ylim(ylim)
+        energy_factor = integral(energyRange) / integral(self._mcHeader['E_range'])
 
-            else:
-                # 1D histogram
+        # core factor
+        core_factor = math.pow(coreMax, 2) / math.pow(self._mcHeader['core_range'][1], 2)
 
-                xlim = _get_ax_lim(hist, axis=0)
+        return self._mcHeader['n_events'] * energy_factor * core_factor
 
-                if np.sum(hist['data']) == 0:
-                    ax.text(
-                        0.5,
-                        0.5,
-                        'EMPTY',
-                        horizontalalignment='center',
-                        verticalalignment='center',
-                        transform=ax.transAxes
-                    )
-                    continue
+    def _validateEnergyRange(self, energyRange):
+        if energyRange is None:
+            return self._mcHeader['E_range']
 
-                x_bins = _get_bins(hist, axis=0)
-                centers = 0.5 * (x_bins[:-1] + x_bins[1:])
-                ax.hist(centers, bins=x_bins, weights=hist['data'])
-                ax.set_xlim(xlim)
+        if not isinstance(energyRange[0], u.Quantity) or not isinstance(energyRange[1], u.Quantity):
+            msg = 'energyRange must be given as u.Quantity in units of energy'
+            self._logger.error(msg)
+            raise TypeError(msg)
 
-            plt.tight_layout()
-            pdfPages.savefig(fig)
-            plt.close()
+        try:
+            return (energyRange[0].to(u.TeV).value, energyRange[1].to(u.TeV).value)
+        except u.core.UnitConversionError:
+            msg = 'energyRange must be in units of energy'
+            self._logger.error(msg)
+            raise TypeError(msg)
 
-        pdfPages.close()
-    # End
+    def _validateCoreMax(self, coreMax):
```

Comment:
Same comment as for the previous function. In addition, in the previous function you made tests and exceptions about the units and here you assume they are OK. Maybe it would be OK in this case if you had had the `@u.quantity_input(coreMax=u.m)` decorator?

Reply:
Yes, for the coreMax we have the decorators to assure the units is fine.
We don't use it for the energyRange because it only works with scalers, not with arrays.

I added the docstrings.

Reply:
And you don't need the decorator also in this function? Or you trust that you always get it "validated" in the calling function?

Reply:
I trust.
The argument should be checked by the public method. This private method is meant to be called by the public ones, so the argument should be validated already.

Reply:
OK!